### DISCUSSION TITLE
Proposal: Allow switch expressions in expression trees

### DISCUSSION THREAD

**Discussion by hazzik** [dotnet/csharplang#4821] [Language Ideas] [answered]
It will be god to be able to use switch expression in expression trees.

```csharp
Expression<Func<int?, string>> i = x => x switch
{
    null => "null",
    1 => "1",
    _ => "default"
};
```

Expression tree do support `Expression.Switch`, so the snipped above could be translated to following expression tree:

```csharp
var parameter = Expression.Parameter(typeof(int), "x");
var expression = Expression.Lambda<Func<int?, string>>(
    Expression.Switch(parameter,
        Expression.Constant("default", typeof(string)),
        Expression.SwitchCase(
            Expression.Constant("null", typeof(string)),
            Expression.Constant(null, typeof(int?))),
        Expression.SwitchCase(
            Expression.Constant("1", typeof(string)),
            Expression.Constant(1, typeof(int?)))),
    parameter);
```

However `Expression.Switch` supports only constant comparison. So, it might be a good idea to translate to a set of `Expression.Condition` instead.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/158

**Reply by hazzik** [reply]
Thanks. #158 is related, but unfortunately is quite different and frankly not so useful for having a productive discussion:

- it is a moving goal post type of issue, because during the lifespan of that discussion lambda expressions in a form of a delegate has received so many features that are not even mentioned in that discussion.
- it would be not practical nor possible to implement all features missing from expression trees in one go.
- switch expressions are not mentioned in that discussion.

**Reply by Joe4evr** [reply]
> * switch expressions are not mentioned in that discussion.

Switch expressions didn't exist yet at the time of that discussion.
> * it is a moving goal post type of issue, because during the lifespan of that discussion lambda expressions in a form of a delegate has received so many features that are not even mentioned in that discussion.

![](https://i.imgur.com/jks6Qqr.png)
The **entire point** of that discussion is to talk about features missing from expression trees, no matter when the feature was made.

**Reply by hazzik** [reply]
> Switch expressions didn't exist yet at the time of that discussion.

Exactly. 

> The entire point of that discussion is to talk about features missing from expression trees, no matter when the feature was made.

There is no point talking about missing features from expression trees in their entirety because it is not actionable. It is much more productive to talk about concrete features that are missing, how to implement them and if they are feasible.

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/discussions/4727

**Reply by hazzik** [reply]
Just to clarify: the main reason this discussion exists is to transition this particular feature into a proposal stage.

**Reply by 333fred** [reply]
To be clear: the cost of adding new things to expression trees is _extremely_ high. We wouldn't do it for just one feature. Is it moving the goalposts? Maybe a bit. But you won't see progress on this piece until you see progress on the whole package.

**Reply by 333fred** [reply]
> Just to clarify: the main reason this discussion exists is to transition this particular feature into a proposal stage.

The proposals already exist, and they're the ones ufcpp linked.

**Reply by hazzik** [reply]
> The proposals already exist, and they're the ones ufcpp linked.

sorry Fred I wrote my last message before I saw the last comment from ufcpp.

> the cost of adding new things to expression trees is extremely high.

Apparently it took me much less time to implement the thing (from zero knowledge of Roslyn source code) than finding a correct way to submit it.

**Reply by 333fred** [reply]
> Apparently it took me much less time to implement the thing (from zero knowledge of Roslyn source code) than finding a correct way to submit it.

Lowering (which you showed an example of for an extremely basic case of a switch expression that has no nested pattern matching) is only half the issue, and it's the easy half. The real problem is that consumers of expression trees will not understand the resulting trees. Sure, the compiler could produce these nodes from an expression tree, but EF isn't going to know how to turn that into SQL and the code will fail at runtime. I'd suggest looking through the linked issues, particularly the second one, and reading the summaries of the issues with evolving the API before suggesting that it's simple.

**Reply by hazzik** [reply]
> The real problem is that consumers of expression trees will not understand the resulting trees.

This is the exact reason I suggested to rewrite into `ConditionalExpression` which is understandable by most Linq providers.

> I'd suggest looking through the linked issues, particularly the second one

I did read it. And it is indeed really hard to implement one solution for all missing tree expression features. 

I’m going to abandon this issue and focus on the linked issue instead.
